micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,514 @@
+//! Contains the DataStore trait and a usable StaticMemoryDataStore.
+//! Implementers of the trait are meant to be written to by DataCollectors (RSDK-6992, RSDK-6994)
+//! and read from by a task that uploads the data to app (RSDK-6995)
+
+use crate::proto::app::data_sync::v1::SensorData;
+use bytes::{BufMut, BytesMut};
+use prost::{length_delimiter_len, EncodeError, Message};
+use ringbuf::{
+    ring_buffer::{RbBase, RbRead, RbWrite},
+    Consumer, LocalRb, Producer, Rb,
+};
+use std::{
+    mem::MaybeUninit,
+    sync::atomic::{AtomicBool, Ordering},
+};
+use thiserror::Error;
+
+use super::data_collector::ResourceMethodKey;
+
+static mut DATA_STORE: [MaybeUninit<u8>; 1024000] = [MaybeUninit::uninit(); 1024000];
+
+#[derive(Error, Debug)]
+pub enum DataStoreError {
+    #[error(transparent)]
+    EncodingError(#[from] EncodeError),
+    #[error("SensorDataTooLarge")]
+    DataTooLarge,
+    #[error("Store already initialized")]
+    DataStoreInitialized,
+    #[error("Data write failure")]
+    DataWriteFailure,
+    #[error("Buffer full")]
+    DataBufferFull(ResourceMethodKey, SensorData),
+    #[error("Current message is malformed")]
+    DataIntegrityError,
+    #[error("Unknown collector key: {0}")]
+    UnknownCollectorKey(ResourceMethodKey),
+    #[error(transparent)]
+    LengthDecodeError(#[from] leb128::read::Error),
+    #[error("unimplemented")]
+    Unimplemented,
+}
+
+lazy_static::lazy_static! {
+    static ref DATA_STORE_INITIALIZED: AtomicBool = AtomicBool::new(false);
+}
+
+pub trait DataStore {
+    /// Store the sensor data message in a region specified by the ResourceMethodKey. To overwrite
+    /// the oldest messages if necessary, pass true for `overwrite_old_data`
+    fn store_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        message: SensorData,
+        overwrite_old_data: bool,
+    ) -> Result<(), DataStoreError>;
+
+    /// Reads the next available message in the store for the given ResourceMethodKey. It should return
+    /// an empty BytesMut with 0 capacity when there are no available messages left.
+    fn read_next_message(
```

Comment:
Can we make the verb part of this method name pair better with `store_message` or its renaming? `store/retrieve`, `push/pull`, `read/write`.